<a href="https://colab.research.google.com/github/RiteshopShrivastava/Python/blob/main/Feature_Engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Univariate feature selection**

In [2]:
# Feature Extraction with Univariate Statistical Tests (Chi-squared for classification)
import pandas as pd
import numpy as np
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
from sklearn.feature_selection import f_classif 
# load data
filename = ('pima-indians-diabetes.data.csv')
column = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = pd.read_csv(filename, names=column)
array = dataframe.values
X= array[:,0:8]
Y = array[:,8]
# feature extraction
model_UV = SelectKBest(score_func =chi2, k=6 )
fit = model_UV.fit(X,Y)
# Summary scores
np.set_printoptions(precision=3)
print(fit.scores_)
#features = fit.transform(X)

#test = SelectKBest(score_func =f_classif, k=4 )
#fit1 = test.fit(X,Y)
#print(fit1.scores_)

[ 111.52  1411.887   17.605   53.108 2175.565  127.669    5.393  181.304]


from above output 'test' (2175.565), 'plas'(1411.887), 'class' (181.304) and 'pedi' (127.669) are most important features

**Recursive Feature Elimination**

In [3]:
# Feature Extraction with RFE
import pandas as pd
import numpy as np
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
# load data
filename = 'pima-indians-diabetes.data.csv'
column = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe_RFE = pd.read_csv(filename,names=column)
array = dataframe_RFE.values
X= array[:, 0:8]
Y = array[:,8]
# feature extraction
#model = LogisticRegression(max_iter=200)                  # put max_itr= 200(more than 100), else will give error
model_rfe = RFE(LogisticRegression(max_iter=200), 3)       # 3 means no of features to be identified
fit = model_rfe.fit(X,Y)
print(fit.n_features_)
print(fit.support_)
print(fit.ranking_)


3
[ True False False False False  True  True False]
[1 2 4 6 5 1 1 3]


From above output we can say,preg,mass,pedi are most important feature

**Feature Importance using Decision Tree**

In [4]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
filename = 'pima-indians-diabetes.data.csv'
column_name = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe_DT = pd.read_csv(filename, names= column_name)
array = dataframe_DT.values
X= array [:,0:8]
Y = array [:, 8]
# Feature extraction
model_DT = DecisionTreeClassifier()
fit = model_DT.fit(X,Y)
print(fit.feature_importances_)

[0.051 0.329 0.097 0.009 0.053 0.215 0.138 0.109]
